### DISCUSSION TITLE
Recommended approach to manage certs and profiles for 2FA enabled Enterprise account

### DISCUSSION THREAD

**Discussion by snelloru** [fastlane/fastlane#20422] [General]
Hi all, thanks for all the hard work in making the Fastlane tool awesome and more useful. I am fairly new to the community and struggling to get the required information. 

We want to automate the management of certificates and profiles for both enterprise and AppStore teams. We have a service account to manage our systems which are part of our CI/CD and this has 2FA enabled as it's mandatory to have 2FA for all new accounts.

I started with Fastlane match to manage certs and profiles and learned that recommended approach is to use app_store_connected_api_key which uses the Apple API key. It's working well for the AppStore accounts but not for enterprise accounts as there is no API key support for enterprise accounts. 

The only solution I see is the FASTLANE_SESSION environment variable through Spaceauth, but this is not a permanent solution and needs manual intervention once in a while. Is this the approach or solution everyone using for your CI/CD using fastlane?

I am wondering, what everyone is doing to manage certificates and profiles for enterprise accounts that are 2FA enabled? I am looking for better solutions without workarounds.

**Comment by Kaspik** [comment]
We did exactly that with the `FASTLANE_SESSION` and regenerating it every 30 days - but because Apple deprecated the Enterprise account lately and won't renew it, we just switched fully to the AppStore one. You might be wasting time here cause your next Enterprise account renewal might not happen too. :)

**Comment by snelloru** [comment]
Thank you for your reply. I had no clue about the news, Apple not renewing
enterprise accounts. Looks like this is only for organizations with less
than 100 employees. We are way more than that, so I believe it's not an
issue.

Was the FASTLANE_SESSION approach reliable? I mean, was it enough to
regularly regenerating the session for every 30 days or it used to fail
intermittently before 30 days? I am afraid, I might introduce another
snowflake in our CI/CD process.

We have many engineers who have access to developer portal/profiles. It's
very common that engineers add their certificates to the profiles and
regenerating them. So, every time, when profiles get regenerated, our
builds fail unless we download all latest profiles into all our build
machines.

I want to solve this problem by using Fastlane Match. I wanted to add a
lane before the build process begins, so that match will download all
latest profiles into servers before the build. But during this process, I
realized it's not easy to completely automate this process because of Apple
2FA.

May I know how did you solve this problem and was it reliable to use it in
our CI/CD automation? We have multiple iOS teams using our CI/CD process.

**Comment by Kaspik** [comment]
We did the download profiles on every CI run to avoid these issues, and that used the session so it was okay.

It was reliable for the 30 days interval, we had Slack reminder every 30 days to do it, but all the CI machines have to be at the same location basically (you can't have one in US and one in Europe).

There is an issue with cookies generated before they expire - https://github.com/fastlane/fastlane/issues/20208

Here is what we did (anonymized) for the Enterprise release on push to master:

```
desc "Builds Alpha target with release configuration, sends it to the Firebase App Distribution, uploads dSYMs"
   lane :build_alpha_target do |options|
      isMaster = options[:master]
      get_certificates(
         team_id: "team_id_here"
      )
      get_provisioning_profile(
         app_identifier: "app_id_here",
         team_id: "team_id_here",
         readonly: true
      )
      increment_build_number(
         build_number: Time.now.to_i
      )
      build_ios_app(
         project: PROJECT_NAME,
         scheme: scheme_here,
         export_method: "enterprise",
         include_bitcode: true,
         include_symbols: true,
         silent: true,
         sdk: DEFAULT_SDK
      )
      sh '../Scripts/upload-symbols -gsp ../Supporting\ Files/Beta/GoogleService-Info.plist -p ios ../Alpha.app.dSYM.zip'
      if isMaster
         sh 'firebase appdistribution:distribute ../Alpha.ipa --app app_id_in_firebase --release-notes "Build created from master" --groups "tester"'
      else
         sh 'firebase appdistribution:distribute ../Alpha.ipa --app app_id_in_firebase --release-notes "Build created manually from PR" --groups "tester"'
      end
   end
```

The `FASTLANE_SESSION` was stored in `.env` file (check `dotenv` here: https://docs.fastlane.tools/best-practices/keys/)

**Comment by snelloru** [comment]
Thank you so much for your help. I checked with my team and we found some old Apple accounts that doesn't have 2FA enabled. I think that makes things more easier. My co-worker was joking that Apple accounts with no 2FA are a commodity now, apparently they are :) LOL
